File: packages/types/src/review.ts
/** Phase 5 v1 wire-contract drafts. Runtime checks live in future service slices.
 * UUID, GitHub IDs, SHAs, timestamps and decimal USD amounts serialize as strings.
 * Numeric values are nonnegative/positive integers as constrained in review.py.
 */
export type GitHubId = string
export type HeadSha = string
export type MoneyUSD = string // Fixed six decimal places; never a JS float.
export type Focus = 'security' | 'performance' | 'quality'
export type Severity = 'CRITICAL' | 'HIGH' | 'MEDIUM' | 'LOW'
export type AnalysisStatus =
  | 'draft' | 'queued' | 'processing' | 'analyzed' | 'partial'
  | 'failed' | 'skipped' | 'superseded' | 'cancelled'
export type PublicationStatus =
  | 'not_requested' | 'preview_ready' | 'publishing' | 'published'
  | 'partially_published' | 'publish_unknown' | 'failed' | 'blocked'
export type ErrorCode =
  | 'validation_failed' | 'access_denied' | 'installation_revoked' | 'stale_head'
  | 'budget_exceeded' | 'diff_unavailable' | 'provider_failed' | 'invalid_findings'
  | 'rate_limited' | 'persistence_failed' | 'publication_failed' | 'publication_unknown'
  | 'cancelled' | 'not_applicable'

export const ANALYSIS_TRANSITIONS = {
  draft: ['queued', 'cancelled'],
  queued: ['processing', 'skipped', 'superseded', 'cancelled'],
  processing: ['queued', 'analyzed', 'partial', 'failed', 'superseded', 'cancelled'],
  analyzed: ['superseded'], partial: ['superseded'],
  failed: [], skipped: [], superseded: [], cancelled: [],
} as const satisfies Record<AnalysisStatus, readonly AnalysisStatus[]>

export const PUBLICATION_TRANSITIONS = {
  not_requested: ['preview_ready', 'blocked'],
  preview_ready: ['publishing', 'blocked'],
  publishing: ['published', 'partially_published', 'publish_unknown', 'failed', 'blocked'],
  published: [],
  partially_published: ['publishing', 'publish_unknown', 'blocked'],
  publish_unknown: ['published', 'partially_published', 'preview_ready', 'blocked'],
  failed: ['preview_ready', 'blocked'], blocked: [],
} as const satisfies Record<PublicationStatus, readonly PublicationStatus[]>

export interface ReviewOwner { user_id: string; installation_id: GitHubId }
export interface ReviewRepository { id: GitHubId; owner_login: string; name: string }
export interface ReviewBudgets {
  max_files: number
  min_changed_lines: number
  max_changed_lines: number
  max_context_lines_per_file: number
  max_chunks: number
  context_window_tokens: number
  max_request_input_tokens: number
  reserved_output_tokens: number
  max_total_input_tokens: number
  max_total_output_tokens: number
  max_findings: number
  max_comments: number
  max_cost_usd: MoneyUSD
  max_owner_daily_cost_usd: MoneyUSD
  token_count_mode: 'provider_exact' | 'conservative_estimate'
}
export interface ReviewPolicy {
  rules_version: number
  enabled: boolean
  focus: Focus[]
  skip_paths: string[]
  provider: 'anthropic'
  model: string
  budgets: ReviewBudgets
  publication_mode: 'preview_only' | 'comment'
}
export interface ReviewIdentity {
  contract_version: '1'
  review_id: string
  owner: ReviewOwner
  repo: ReviewRepository
  pr_number: number
  head_sha: HeadSha
  policy: ReviewPolicy
}
export interface FindingLocation { path: string; line: number; side: 'LEFT' | 'RIGHT' }
export interface ReviewFinding {
  id: string
  fingerprint: string
  focus: Focus
  severity: Severity
  location: FindingLocation
  description: string
  suggestion: string
  evidence: string
}
export interface ReviewError {
  code: ErrorCode
  message: string
  retryable: boolean
  request_id: string
  retry_after_seconds: number | null
}
export interface ReviewHTTPError { detail: ReviewError }
export interface ReviewCoverage {
  files_total: number
  files_reviewed: number
  changed_lines_total: number
  changed_lines_reviewed: number
  gaps: string[]
}
export interface ReviewUsage {
  input_tokens: number
  output_tokens: number
  cost_usd: MoneyUSD
  token_count_mode: 'provider_exact' | 'conservative_estimate'
}
export interface AnalysisDraft { status: 'draft' | 'queued' | 'processing' }
export interface AnalysisResult {
  status: 'analyzed' | 'partial'
  summary: string
  findings: ReviewFinding[]
  coverage: ReviewCoverage
  usage: ReviewUsage
}
export interface AnalysisStopped {
  status: 'failed' | 'skipped' | 'superseded' | 'cancelled'
  error: ReviewError
}
export type Analysis = AnalysisDraft | AnalysisResult | AnalysisStopped
export interface PublicationDraft { status: 'not_requested' | 'preview_ready' }
export interface PublicationAttempt {
  status: 'publishing'
  attempt_id: string
  head_sha: HeadSha
  event: 'COMMENT'
}
export interface PublicationReceipt {
  status: 'published' | 'partially_published'
  attempt_id: string
  head_sha: HeadSha
  event: 'COMMENT'
  github_review_ids: GitHubId[]
  confirmed_at: string
}
export interface PublicationUnknown {
  status: 'publish_unknown'
  attempt_id: string
  head_sha: HeadSha
  error: ReviewError
}
export interface PublicationStopped { status: 'failed' | 'blocked'; error: ReviewError }
export type Publication =
  | PublicationDraft | PublicationAttempt | PublicationReceipt
  | PublicationUnknown | PublicationStopped
export interface ReviewRecord extends ReviewIdentity {
  analysis: Analysis
  publication: Publication
}

File: apps/api/src/plutolab_api/schemas/review.py
"""Phase 5 v1 contract drafts; no routes, persistence, workers, or external calls.

HTTP errors retain FastAPI's ``detail`` envelope. Analysis and publication are
separate tagged unions: an analyzed result is never proof of publication.
Transition maps describe permitted edges; enforcement belongs to later slices.
"""

from datetime import datetime
from typing import Annotated, Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

GitHubId = Annotated[str, Field(pattern=r"^[1-9][0-9]*$")]
HeadSha = Annotated[str, Field(pattern=r"^([0-9a-f]{40}|[0-9a-f]{64})$")]
PositiveInt = Annotated[int, Field(strict=True, gt=0)]
NonNegativeInt = Annotated[int, Field(strict=True, ge=0)]
MoneyUSD = Annotated[str, Field(pattern=r"^(0|[1-9][0-9]*)\.[0-9]{6}$")]
Focus = Literal["security", "performance", "quality"]
Severity = Literal["CRITICAL", "HIGH", "MEDIUM", "LOW"]
AnalysisStatus = Literal[
    "draft",
    "queued",
    "processing",
    "analyzed",
    "partial",
    "failed",
    "skipped",
    "superseded",
    "cancelled",
]
PublicationStatus = Literal[
    "not_requested",
    "preview_ready",
    "publishing",
    "published",
    "partially_published",
    "publish_unknown",
    "failed",
    "blocked",
]
ErrorCode = Literal[
    "validation_failed",
    "access_denied",
    "installation_revoked",
    "stale_head",
    "budget_exceeded",
    "diff_unavailable",
    "provider_failed",
    "invalid_findings",
    "rate_limited",
    "persistence_failed",
    "publication_failed",
    "publication_unknown",
    "cancelled",
    "not_applicable",
]

ANALYSIS_TRANSITIONS: dict[AnalysisStatus, tuple[AnalysisStatus, ...]] = {
    "draft": ("queued", "cancelled"),
    "queued": ("processing", "skipped", "superseded", "cancelled"),
    "processing": ("queued", "analyzed", "partial", "failed", "superseded", "cancelled"),
    "analyzed": ("superseded",),
    "partial": ("superseded",),
    "failed": (),
    "skipped": (),
    "superseded": (),
    "cancelled": (),
}
PUBLICATION_TRANSITIONS: dict[PublicationStatus, tuple[PublicationStatus, ...]] = {
    "not_requested": ("preview_ready", "blocked"),
    "preview_ready": ("publishing", "blocked"),
    "publishing": ("published", "partially_published", "publish_unknown", "failed", "blocked"),
    "published": (),
    "partially_published": ("publishing", "publish_unknown", "blocked"),
    "publish_unknown": ("published", "partially_published", "preview_ready", "blocked"),
    "failed": ("preview_ready", "blocked"),
    "blocked": (),
}


class ReviewContract(BaseModel):
    model_config = ConfigDict(extra="forbid")


class ReviewOwner(ReviewContract):
    """Server-derived owner; never accepted as client authorization evidence."""

    user_id: UUID
    installation_id: GitHubId


class ReviewRepository(ReviewContract):
    id: GitHubId
    owner_login: str = Field(min_length=1, max_length=100)
    name: str = Field(min_length=1, max_length=100)


class ReviewBudgets(ReviewContract):
    """Required policy inputs: no implicit spend or hard-coded provider window."""

    max_files: PositiveInt
    min_changed_lines: NonNegativeInt
    max_changed_lines: PositiveInt
    max_context_lines_per_file: NonNegativeInt
    max_chunks: PositiveInt
    context_window_tokens: PositiveInt
    max_request_input_tokens: PositiveInt
    reserved_output_tokens: PositiveInt
    max_total_input_tokens: PositiveInt
    max_total_output_tokens: PositiveInt
    max_findings: PositiveInt
    max_comments: PositiveInt
    max_cost_usd: MoneyUSD
    max_owner_daily_cost_usd: MoneyUSD
    token_count_mode: Literal["provider_exact", "conservative_estimate"]


class ReviewPolicy(ReviewContract):
    rules_version: PositiveInt
    enabled: bool = False
    focus: list[Focus] = Field(min_length=1, max_length=3)
    skip_paths: list[str] = Field(max_length=100)
    provider: Literal["anthropic"]
    model: str = Field(min_length=1, max_length=200)
    budgets: ReviewBudgets
    publication_mode: Literal["preview_only", "comment"] = "preview_only"


class ReviewIdentity(ReviewContract):
    contract_version: Literal["1"] = "1"
    review_id: UUID
    owner: ReviewOwner
    repo: ReviewRepository
    pr_number: PositiveInt
    head_sha: HeadSha
    policy: ReviewPolicy


class FindingLocation(ReviewContract):
    """Repository-relative path and new/old diff line; never a URL or command."""

    path: str = Field(min_length=1, max_length=4096)
    line: PositiveInt
    side: Literal["LEFT", "RIGHT"]


class VerifiedDiffPosition(ReviewContract):
    path: str = Field(min_length=1, max_length=4096)
    line: PositiveInt
    side: Literal["LEFT", "RIGHT"]
    position: PositiveInt
    comment_context: str = Field(min_length=1, max_length=20_000)


class VerifiedDiffSnapshot(ReviewContract):
    head_sha: HeadSha
    findings: dict[str, VerifiedDiffPosition] = Field(max_length=100)


class ReviewFinding(ReviewContract):
    id: UUID
    fingerprint: str = Field(min_length=1, max_length=128)
    focus: Focus
    severity: Severity
    location: FindingLocation
    description: str = Field(min_length=1, max_length=2000)
    suggestion: str = Field(min_length=1, max_length=4000)
    evidence: str = Field(min_length=1, max_length=4000)


class ReviewError(ReviewContract):
    code: ErrorCode
    message: str = Field(min_length=1, max_length=1000)
    retryable: bool
    request_id: UUID
    retry_after_seconds: NonNegativeInt | None = None


class ReviewHTTPError(ReviewContract):
    """Review-specific HTTPException detail; no new global response envelope."""

    detail: ReviewError


class ReviewCoverage(ReviewContract):
    files_total: NonNegativeInt
    files_reviewed: NonNegativeInt
    changed_lines_total: NonNegativeInt
    changed_lines_reviewed: NonNegativeInt
    gaps: list[str] = Field(max_length=100)


class ReviewUsage(ReviewContract):
    input_tokens: NonNegativeInt
    output_tokens: NonNegativeInt
    cost_usd: MoneyUSD
    token_count_mode: Literal["provider_exact", "conservative_estimate"]


class AnalysisDraft(ReviewContract):
    status: Literal["draft", "queued", "processing"]


class AnalysisResult(ReviewContract):
    status: Literal["analyzed", "partial"]
    summary: str = Field(min_length=1, max_length=10000)
    findings: list[ReviewFinding]
    coverage: ReviewCoverage
    usage: ReviewUsage
    verified_diff: VerifiedDiffSnapshot | None = None


class AnalysisStopped(ReviewContract):
    status: Literal["failed", "skipped", "superseded", "cancelled"]
    error: ReviewError


Analysis = Annotated[
    AnalysisDraft | AnalysisResult | AnalysisStopped, Field(discriminator="status")
]


class PublicationDraft(ReviewContract):
    status: Literal["not_requested", "preview_ready"]


class PublicationAttempt(ReviewContract):
    status: Literal["publishing"]
    attempt_id: UUID
    head_sha: HeadSha
    event: Literal["COMMENT"] = "COMMENT"


class PublicationReceipt(ReviewContract):
    status: Literal["published", "partially_published"]
    attempt_id: UUID
    head_sha: HeadSha
    event: Literal["COMMENT"] = "COMMENT"
    github_review_ids: list[GitHubId] = Field(min_length=1)
    confirmed_at: datetime
    fallback_reason: Literal["diff_position_mismatch"] | None = None


class PublicationUnknown(ReviewContract):
    status: Literal["publish_unknown"]
    attempt_id: UUID
    head_sha: HeadSha
    error: ReviewError


class PublicationStopped(ReviewContract):
    status: Literal["failed", "blocked"]
    error: ReviewError


Publication = Annotated[
    PublicationDraft
    | PublicationAttempt
    | PublicationReceipt
    | PublicationUnknown
    | PublicationStopped,
    Field(discriminator="status"),
]


class ReviewRecord(ReviewIdentity):
    analysis: Analysis
    publication: Publication
